File: java/src/test/java/accounting/patterns/MonetaryAmountTest.java
package accounting.patterns;


import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.Currency;

import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;


public class MonetaryAmountTest {
    @Test
    public void differentIfAmountsDiffer() throws Exception {
        final MonetaryAmount monetaryAmount1 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(1));
        final MonetaryAmount monetaryAmount2 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(2));


        assertFalse(monetaryAmount1.equals(monetaryAmount2));
    }

    @Test
    public void differentIfCurrenciesDiffer() throws Exception {
        final MonetaryAmount monetaryAmount1 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(1));
        final MonetaryAmount monetaryAmount2 = new MonetaryAmount(Currency.getInstance("GBP"), new BigDecimal(1));


        assertFalse(monetaryAmount1.equals(monetaryAmount2));
    }

    @Test
    public void sameIfCurrencyAndValueMatch() throws Exception {
        final MonetaryAmount monetaryAmount1 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(1));
        final MonetaryAmount monetaryAmount2 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(1));


        assertTrue(monetaryAmount1.equals(monetaryAmount2));
    }

    @Test
    public void hashcodeVariesByCurrencyAndValue() throws Exception {
        final MonetaryAmount monetaryAmount1 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(1));
        final MonetaryAmount monetaryAmount2 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(1));
        final MonetaryAmount monetaryAmount3 = new MonetaryAmount(Currency.getInstance("GBP"), new BigDecimal(1));
        final MonetaryAmount monetaryAmount4 = new MonetaryAmount(Currency.getInstance("USD"), new BigDecimal(2));


        assertTrue(monetaryAmount1.hashCode() == monetaryAmount2.hashCode());

        assertFalse(monetaryAmount1.hashCode() == monetaryAmount3.hashCode());
        assertFalse(monetaryAmount1.hashCode() == monetaryAmount4.hashCode());
    }
}

File: src/test/typescript/accounting-pattern/MonetaryAmountTest.ts
import { MonetaryAmount } from "../../main/typescript/accounting-pattern/MonetaryAmount";

describe("MonetaryAmount", () => {
    it("should be different if amounts differ", () => {
        const monetaryAmount1 = new MonetaryAmount("USD", 1);
        const monetaryAmount2 = new MonetaryAmount("USD", 2);

        expect(monetaryAmount1.equals(monetaryAmount2)).toBe(false);
    });

    it("should be different if currencies differ", () => {
        const monetaryAmount1 = new MonetaryAmount("USD", 1);
        const monetaryAmount2 = new MonetaryAmount("GBP", 1);

        expect(monetaryAmount1.equals(monetaryAmount2)).toBe(false);
    });

    it("should be the same if currency and value match", () => {
        const monetaryAmount1 = new MonetaryAmount("USD", 1);
        const monetaryAmount2 = new MonetaryAmount("USD", 1);

        expect(monetaryAmount1.equals(monetaryAmount2)).toBe(true);
    });

    it("should have a hashcode that varies by currency and value", () => {
        const monetaryAmount1 = new MonetaryAmount("USD", 1);
        const monetaryAmount2 = new MonetaryAmount("USD", 1);
        const monetaryAmount3 = new MonetaryAmount("GBP", 1);
        const monetaryAmount4 = new MonetaryAmount("USD", 2);

        expect(monetaryAmount1.hashCode()).toBe(monetaryAmount2.hashCode());
        expect(monetaryAmount1.hashCode()).not.toBe(monetaryAmount3.hashCode());
        expect(monetaryAmount1.hashCode()).not.toBe(monetaryAmount4.hashCode());
    });
});
